common/timer.go

```diff
@@ -152,74 +185,114 @@ func (tstore *TimerStore) setNextTimeout(ctx sdk.Context, which types.TimerType,
+func (tstore *TimerStore) delTimer(ctx sdk.Context, which types.TimerType, value uint64, key []byte) {
 	store := tstore.getStoreTimer(ctx, which)
-	store.Delete(types.EncodeKey(value))
+	timerKey := types.EncodeBlockAndKey(value, key)
+	if !store.Has(timerKey) {
+		panic(fmt.Sprintf("delTimer which %d block %d key %v: no such timer", which, value, key))
+	}
+	store.Delete(timerKey)
 }
 
-func (tstore *TimerStore) AddTimerByBlockHeight(ctx sdk.Context, block uint64, data string) {
-	tstore.addTimer(ctx, types.BlockHeight, block, data)
+// AddTimerByBlockHeight adds a new timer to expire on a given block height.
+// If a timer for that <block, key> tuple exists, it will be overridden.
+func (tstore *TimerStore) AddTimerByBlockHeight(ctx sdk.Context, block uint64, key []byte, data []byte) {
+	if block <= uint64(ctx.BlockHeight()) {
+		panic(fmt.Sprintf("timer expiry block %d smaller than ctx block %d",
+			block, uint64(ctx.BlockHeight())))
+	}
+	tstore.addTimer(ctx, types.BlockHeight, block, key, data)
 }
 
-func (tstore *TimerStore) AddTimerByBlockTime(ctx sdk.Context, timestamp uint64, data string) {
-	tstore.addTimer(ctx, types.BlockTime, timestamp, data)
+// AddTimerByBlockTime adds a new timer to expire on a future block with the given timestamp.
+// If a timer for that <timestamp, key> tuple exists, it will be overridden.
+func (tstore *TimerStore) AddTimerByBlockTime(ctx sdk.Context, timestamp uint64, key []byte, data []byte) {
+	if timestamp <= uint64(ctx.BlockTime().UTC().Unix()) {
+		panic(fmt.Sprintf("timer expiry time %d smaller than ctx time %d",
+			timestamp, uint64(ctx.BlockTime().UTC().Unix())))
+	}
+	tstore.addTimer(ctx, types.BlockTime, timestamp, key, data)
 }
 
-type timerTuple struct {
-	value uint64
-	data  string
+// DelTimerByBlockHeight removes an eximsting timer for the <block, key> tuple.
+func (tstore *TimerStore) DelTimerByBlockHeight(ctx sdk.Context, block uint64, key []byte) {
+	tstore.delTimer(ctx, types.BlockHeight, block, key)
 }
 
-func (tstore *TimerStore) tickValue(ctx sdk.Context, which types.TimerType, tickValue uint64) {
-	nextValue := tstore.getNextTimeout(ctx, which)
-	if tickValue < nextValue {
-		return
+// DelTimerByBlockHeight removes an eximsting timer for the <timetamp, key> tuple.
+func (tstore *TimerStore) DelTimerByBlockTime(ctx sdk.Context, timestamp uint64, key []byte) {
+	tstore.delTimer(ctx, types.BlockTime, timestamp, key)
+}
+
+func (tstore *TimerStore) DumpTimers(ctx sdk.Context, which types.TimerType) string {
+	store := tstore.getStoreTimer(ctx, which)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	var b strings.Builder
+
+	for ; iterator.Valid(); iterator.Next() {
+		value, key := types.DecodeBlockAndKey(iterator.Key())
+		b.WriteString(fmt.Sprintf("block: %d key %v\n", value, key))
 	}
 
+	return b.String()
+}
+
+func (tstore *TimerStore) getFrontTimer(ctx sdk.Context, which types.TimerType) (uint64, []byte, []byte) {
```

Comment:
According to your example from the documentation above, an object can have multiple timers of different types. In your example, a package module's object has a blockHeight timer that expires on block 150 and three additional timers that expire on block 180. Also, it has a blockTime timer that never expires (maxUint64).

What is the hierarchy between them? According to getFrontTimer's implementation, you get the latest timer of a certain type. But since all the blockHeight timers will expire before the BlockTime timer expire, is it not correct returning them first? why do you differentiate between the timer types? you can convert the blockTime timers expiry field to blocks and have one list of timers.

Reply:
i conversion of time to block and vise versa is a problematic area, i think we want to avoid that

Reply:
> According to your example from the documentation above, an object can have multiple timers of different types. 

Actually, modules have timers, usually related to objects they manage ("objects" themselves do not have timers). Also, note: "multiple types" --> two types: either "block-height" or "block-time".

> In your example, a package module's object has a blockHeight timer that expires on block 150 and three additional timers that expire on block 180. Also, it has a blockTime timer that never expires (maxUint64).

There is no timer of type BlockTime, hence the "next" timeout there is infinity.
There are three timers of type BlockHeight: "first", "second", "third" (both keys and data). They differ in their key - each key usually "represents" a distinct object, as decided by the module. If the same key (and expiry block/time) is used for two different objects, then one's timer will overwrite the other's. 

> According to getFrontTimer's implementation, you get the latest timer of a certain type. 

Not the latest - but the earliest, i.e. the one coming up next.
Usually you can get it by looking at the record for the "next" timeout, but during the loop in `tickValue()` we can't reliably find the new "next" timeout (see comment there), so we need this function.

> But since all the blockHeight timers will expire before the BlockTime timer expire, is it not correct returning them first? 

The front - as in earliest - is returned. See the comment in the function.
There are no timers or type BlockTime in the example.
Anyway, the two types of timers are managed separately.

> why do you differentiate between the timer types? you can convert the blockTime timers expiry field to blocks and have one list of timers.

Not possible to convert the BlockTime expiry to blocks: BlockTime based expiry means "expire upon the first block whose timestamp is not earlier than the expiry time". We can't predict exactly how many blocks in the future it will happen.

Splitting into two types simplifies the code and allows to reuse most of the logic except for the calculation of expiry conditions.

Reply:
> i conversion of time to block and vise versa is a problematic area, i think we want to avoid that

why is it problematic?
the following definition is simple and deterministic: "time-based expiry means to expire upon the first block whose timestamp is not earlier than the expiry time".
we already do it with subscriptions.